### PR TITLE
Reading PDB constants

### PR DESCRIPTION
**Pull request by tmat**
The ISymUnmanagedConstant.GetValue returns constant values whose types don't exactly match their signatures in metadata. We need to compensate for that. 

Updates PdbToXml and EEs to handle the constant values correctly and report diagnostics when PDB contains invalid data.

Adds tests to cover all possible kinds of constant types.

### CODE REVIEW

**Comment by tmat** [comment]
@KevinH-MS

File: src/Test/PdbUtilities/Pdb/PdbToXml.cs [hunk 1]
```diff
@@ -964,47 +965,96 @@ private void WriteLocalsHelper(ISymUnmanagedScope scope, Dictionary<int, Immutab
...
-                {
-                    _writer.WriteAttributeString("name", c.GetName());
+                _writer.WriteAttributeString("name", name);
 
-                    object value = c.GetValue();
-                    string typeName = value.GetType().Name;
+                if (value is int &&
+                    (int)value == 0 &&
+                    (signature[0] == (byte)ConstantTypeCode.NullReference ||
+                     signature[0] == (int)SignatureTypeCode.Object ||
```
Comment (cston) [inline src/Test/PdbUtilities/Pdb/PdbToXml.cs]: Perhaps extract signature[0] to a local.

File: src/Compilers/VisualBasic/Test/Emit/PDB/PDBConstLocalTests.vb [hunk 2]
```diff
@@ -169,51 +169,35 @@ End Class
...
-                 Diagnostic(ERRID.WRN_UnusedLocalConst, "dec").WithArguments("dec"),
-                 Diagnostic(ERRID.WRN_UnusedLocalConst, "dt").WithArguments("dt"))
-
-            Dim actual As XElement = GetPdbXml(compilation, "C.M")
-            Dim invariantStr = actual.ToString()
-            invariantStr = invariantStr.Replace("-3,402823E+38", "-3.402823E+38")
-            invariantStr = invariantStr.Replace("1,79769313486232E+308", "1.79769313486232E+308")
-            invariantStr = invariantStr.Replace("2,98187743664301E-266", "2.98187743664301E-266")
-            invariantStr = invariantStr.Replace("value=""1,5""", "value=""1.5""")
-            Dim expected = <symbols>
```
Comment (cston) [inline src/Compilers/VisualBasic/Test/Emit/PDB/PDBConstLocalTests.vb]: Why do the values have comma decimal separator? Isn't the reader using InvariantCulture?

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/PDB/PdbHelpers.cs [hunk 3]
```diff
@@ -51,6 +53,185 @@ internal static void GetAllScopes(this ISymUnmanagedMethod method, ArrayBuilder<
...
+
+                    return ConstantValue.Create((short)symValue);
+
+                case SpecialType.System_Char:
+                    if (!(symValue is ushort))
+                    {
+                        return ConstantValue.Bad;
+                    }
+
+                    return ConstantValue.Create((char)(ushort)symValue);
```
Comment (cston) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/PDB/PdbHelpers.cs]: Any chance symValue is out of range?

File: src/ExpressionEvaluator/VisualBasic/Source/ExpressionCompiler/Symbols/EELocalConstantSymbol.vb [hunk 4]
```diff
@@ -1,6 +1,7 @@
 ﻿' Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
 Imports System.Collections.Immutable
+Imports System.Threading
```
Comment (cston) [inline src/ExpressionEvaluator/VisualBasic/Source/ExpressionCompiler/Symbols/EELocalConstantSymbol.vb]: Unnecessary Imports.

**Comment by cston** [comment]
LGTM

File: src/Compilers/VisualBasic/Test/Emit/PDB/PDBConstLocalTests.vb (hunk 2 above)
Comment (tmat) [inline src/Compilers/VisualBasic/Test/Emit/PDB/PDBConstLocalTests.vb]: Apparently it didn't. Now it does.

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/PDB/PdbHelpers.cs (hunk 3 above)
Comment (tmat) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/PDB/PdbHelpers.cs]: char and ushort have the same range.

File: src/Compilers/CSharp/Portable/CSharpResources.resx [hunk 5]
```diff
@@ -4504,6 +4504,9 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
   <data name="ERR_CantReadRulesetFile" xml:space="preserve">
     <value>Error reading ruleset file {0} - {1}</value>
   </data>
+  <data name="ERR_BadPdbData" xml:space="preserve">
+    <value>Error reading debug information for '{0}'</value>
```
Comment (KevinH-MS) [inline src/Compilers/CSharp/Portable/CSharpResources.resx]: We typically have a period at the end of diagnostic messages.  Analyzer messages above should probably also have periods...

File: src/Compilers/VisualBasic/Portable/VBResources.resx [hunk 6]
```diff
@@ -5255,6 +5255,9 @@
   <data name="ERR_LanguageVersion" xml:space="preserve">
     <value>Visual Basic {0} does not support {1}.</value>
   </data>
+  <data name="ERR_BadPdbData" xml:space="preserve">
+    <value>Error reading debug information for '{0}'</value>
```
Comment (KevinH-MS) [inline src/Compilers/VisualBasic/Portable/VBResources.resx]: period

File: src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/EvaluationContext.cs [hunk 7]
```diff
@@ -476,30 +476,38 @@ internal CompilationContext CreateCompilationContext(CSharpSyntaxNode syntax)
...
+                    var signature = constant.GetSignature();
 
-                var constantValue = ReinterpretConstantValue(constant.Value, type.SpecialType);
-                builder.Add(new EELocalConstantSymbol(method, constant.Name, type, constantValue));
+                    var info = metadataDecoder.GetLocalInfo(signature);
+                    Debug.Assert(!info.IsByRef);
+                    Debug.Assert(!info.IsPinned);
+                    var type = info.Type;
+
+                    var constantValue = PdbHelpers.GetConstantValue(type.EnumUnderlyingType(), rawValue);
```
Comment (KevinH-MS) [inline src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/EvaluationContext.cs]: Won't type.EnumUnderlyingType return null for things that aren't enums?

Comment (KevinH-MS) [inline src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/EvaluationContext.cs]: I see that it's an extension that does something different than the instance method... :-S

Please consider renaming to "EnumUnderlyingTypeOrSelf" (as in VB)

File: src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/EvaluationContext.cs [hunk 8]
```diff
@@ -476,30 +476,38 @@ internal CompilationContext CreateCompilationContext(CSharpSyntaxNode syntax)
...
+                    if (dynamicLocalConstantMap != null && dynamicLocalConstantMap.TryGetValue(name, out dynamicFlags))
+                    {
+                        type = DynamicTypeDecoder.TransformTypeWithoutCustomModifierFlags(
+                            type,
+                            containingAssembly,
+                            RefKind.None,
+                            dynamicFlags);
+                    }
+
+                    builder.Add(new EELocalConstantSymbol(method, name, type, constantValue));
```
Comment (KevinH-MS) [inline src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/EvaluationContext.cs]: What if constantValue.IsBad()?

File: src/Compilers/CSharp/Portable/CSharpResources.resx (hunk 5 above)
Comment (tmat) [inline src/Compilers/CSharp/Portable/CSharpResources.resx]: Presence of period is not consistent at all. Feel free to file an issue.

**Comment by KevinH-MS** [comment]
LGTM

**Pull request merged**